File: src/arch/components/rigid_body.c
#include "components/rigid_body.h"
#include "components/transform.h"

#include "core/log.h"

#include <ode/ode.h>

void registerRigidBody(Scene *scene, UUID entity)
{
	RigidBodyComponent *body = sceneGetComponentFromEntity(
		scene,
		entity,
		idFromName("rigid_body"));
	CollisionComponent *coll = sceneGetComponentFromEntity(
		scene,
		entity,
		idFromName("collision"));
	TransformComponent *trans = sceneGetComponentFromEntity(
		scene,
		entity,
		idFromName("transform"));

	if (kmQuaternionLengthSq(&trans->globalRotation) == 0.0f)
	{
		LOG("ERROR: Rotation with a magnitude of 0 on entity: %s\n",
			entity.string);
		ASSERT(false);
	}

	body->bodyID = dBodyCreate(scene->physicsWorld);
	body->spaceID = dSimpleSpaceCreate(scene->physicsSpace);

	if (coll)
	{
		createCollisionGeoms(scene, trans, body, coll);
	}

	// update all the other information about the rigidbody
	updateRigidBody(scene, coll, body, trans);
}

internal
void updateCollisionGeom(
	Scene *scene,
	UUID entity,
	TransformComponent *bodyTrans,
	TransformComponent *trans,
	CollisionTreeNodeComponent *node)
{
	kmVec3 pos, scal;
	kmQuaternion rot;
	tGetInverseGlobalTransform(bodyTrans, &pos, &rot, &scal);
	kmVec3Add(&pos, &trans->globalPosition, &pos);
	kmQuaternionMultiply(&rot, &rot, &trans->globalRotation);
	kmVec3Mul(&scal, &scal, &trans->globalScale);

	dGeomSetOffsetPosition(
		node->geomID,
		pos.x,
		pos.y,
		pos.z);

	dReal quat[4] = {};
	quat[0] = rot.w;
	quat[1] = rot.x;
	quat[2] = rot.y;
	quat[3] = rot.z;

	dGeomSetOffsetQuaternion(node->geomID, quat);

	real32 maxScale = MAX(
	MAX(trans->globalScale.x,
		trans->globalScale.y),
	trans->globalScale.z);

	BoxComponent *box = NULL;
	SphereComponent *sphere = NULL;
	CapsuleComponent *capsule = NULL;

	switch (node->type)
	{
		case COLLISION_GEOM_TYPE_BOX:
			box = sceneGetComponentFromEntity(scene, entity, idFromName("box"));
			dGeomBoxSetLengths(
				node->geomID,
				box->bounds.x * trans->globalScale.x * 2,
				box->bounds.y * trans->globalScale.y * 2,
				box->bounds.z * trans->globalScale.z * 2);
			break;
		case COLLISION_GEOM_TYPE_SPHERE:
			sphere = sceneGetComponentFromEntity(
				scene,
				entity,
				idFromName("sphere"));
			dGeomSphereSetRadius(node->geomID, sphere->radius * maxScale);
			break;
		case COLLISION_GEOM_TYPE_CAPSULE:
			capsule = sceneGetComponentFromEntity(
				scene,
				entity,
				idFromName("capsule"));
			dGeomCapsuleSetParams(
				node->geomID,
				capsule->radius * maxScale,
				capsule->length * maxScale);
			break;
		default:
			break;
	}
}

internal
void updateCollisionGeoms(
	Scene *scene,
	TransformComponent *bodyTrans,
	CollisionComponent *coll)
{
	CollisionTreeNodeComponent *node = 0;
	TransformComponent *trans = 0;
	UUID collisionTreeNodeComponentID = idFromName("collision_tree_node");
	UUID transformComponentID = idFromName("transform");

	// Walk the tree of collision geometry
	for (UUID currentCollider = coll->collisionTree;
		 strcmp(currentCollider.string, "");
		 currentCollider = node->nextCollider)
	{
		trans = sceneGetComponentFromEntity(
			scene,
			currentCollider,
			transformComponentID);

		node = sceneGetComponentFromEntity(
			scene,
			currentCollider,
			collisionTreeNodeComponentID);

		// Add each piece of collision geometry as a different geom
		updateCollisionGeom(scene, currentCollider, bodyTrans, trans, node);
	}
}

internal
void createCollisionGeom(
	Scene *scene,
	UUID entity,
	TransformComponent *bodyTrans,
	RigidBodyComponent *body)
{
	CollisionTreeNodeComponent *node = sceneGetComponentFromEntity(
		scene,
		entity,
		idFromName("collision_tree_node"));
	TransformComponent *trans = sceneGetComponentFromEntity(
		scene,
		entity,
		idFromName("transform"));

	ASSERT(node && "Collision tree pointed to a node with no node structure");

	if (trans->dirty)
	{
		applyParentTransform(scene, trans);
	}

	real32 maxScale = MAX(
		MAX(trans->globalScale.x,
			trans->globalScale.y),
		trans->globalScale.z);

	switch (node->type)
	{
	case COLLISION_GEOM_TYPE_BOX:
	{
		BoxComponent *box = sceneGetComponentFromEntity(
			scene,
			entity,
			idFromName("box"));

		node->geomID = dCreateBox(
			body->spaceID,
			box->bounds.x * trans->globalScale.x * 2,
			box->bounds.y * trans->globalScale.y * 2,
			box->bounds.z * trans->globalScale.z * 2);
		dGeomSetQuaternion(node->geomID, (dReal*)&trans->globalRotation);
	} break;
	case COLLISION_GEOM_TYPE_SPHERE:
	{
		SphereComponent *sphere = sceneGetComponentFromEntity(
			scene,
			entity,
			idFromName("sphere"));

		node->geomID = dCreateSphere(body->spaceID, sphere->radius * maxScale);
	} break;
	case COLLISION_GEOM_TYPE_CAPSULE:
	{
		CapsuleComponent *capsule = sceneGetComponentFromEntity(
			scene,
			entity,
			idFromName("capsule"));

		node->geomID = dCreateCapsule(
			body->spaceID,
			capsule->radius * maxScale,
			capsule->length * maxScale);
		dGeomSetQuaternion(node->geomID, (dReal*)&trans->globalRotation);
	} break;
	default:
	{
		ASSERT(false && "Unable to determine geometry type for entity");
	} break;
	}

	dGeomSetBody(node->geomID, body->bodyID);

	// TODO: Store the index in the component data table, instead of the uuid
	void *userData = calloc(1, sizeof(UUID));
	memcpy(userData, &entity, sizeof(UUID));
	dGeomSetData(node->geomID, userData);

	updateCollisionGeom(scene, entity, bodyTrans, trans, node);
}

void createCollisionGeoms(
	Scene *scene,
	TransformComponent *bodyTrans,
	RigidBodyComponent *body,
	CollisionComponent *coll)
{
	CollisionTreeNodeComponent *node = 0;
	UUID collisionTreeNodeID = idFromName("collision_tree_node");

	// Walk the tree of collision geometry
	for (UUID currentCollider = coll->collisionTree;
		 strcmp(currentCollider.string, "");
		 currentCollider = node->nextCollider)
	{
		// Add each piece of collision geometry as a different geom
		createCollisionGeom(scene, currentCollider, bodyTrans, body);

		node = sceneGetComponentFromEntity(
			scene,
			currentCollider,
			collisionTreeNodeID);
	}
}

void addForce(RigidBodyComponent *body, kmVec3 *force, kmVec3 *position)
{
	dBodyAddForceAtRelPos(
		body->bodyID,
		force->x,
		force->y,
		force->z,
		position->x,
		position->y,
		position->z);
}

void addTorque(RigidBodyComponent *body, kmVec3 *torque)
{
	dBodyAddTorque(body->bodyID, torque->x, torque->y, torque->z);
}

void setForce(RigidBodyComponent *body, kmVec3 *force)
{
	dBodySetForce(body->bodyID, force->x, force->y, force->z);
}

void setTorque(RigidBodyComponent *body, kmVec3 *torque)
{
	dBodySetTorque(body->bodyID, torque->x, torque->y, torque->z);
}

kmVec3 getForce(RigidBodyComponent *body)
{
	const dReal *forceValues = dBodyGetForce(body->bodyID);

	kmVec3 force;
	kmVec3Fill(&force, forceValues[0], forceValues[1], forceValues[2]);

	return force;
}

kmVec3 getTorque(RigidBodyComponent *body)
{
	const dReal *torqueValues = dBodyGetTorque(body->bodyID);

	kmVec3 torque;
	kmVec3Fill(&torque, torqueValues[0], torqueValues[1], torqueValues[2]);

	return torque;
}

void updateRigidBody(
	Scene *scene,
	CollisionComponent *coll,
	RigidBodyComponent *body,
	TransformComponent *trans)
{
	dMass mass = {};

	switch (body->inertiaType)
	{
	case MOMENT_OF_INERTIA_USER:
	{
		dMassSetParameters(
			&mass,
			body->mass,
			body->centerOfMass.x,
			body->centerOfMass.y,
			body->centerOfMass.z,
			body->moiParams[0],
			body->moiParams[1],
			body->moiParams[2],
			body->moiParams[3],
			body->moiParams[4],
			body->moiParams[5]);
	} break;
	case MOMENT_OF_INERTIA_CUBE:
	{
		dMassSetBoxTotal(
			&mass,
			body->mass,
			body->moiParams[0] * 2,
			body->moiParams[1] * 2,
			body->moiParams[2] * 2);
	} break;
	case MOMENT_OF_INERTIA_SPHERE:
	{
		dMassSetSphereTotal(&mass, body->mass, body->moiParams[0]);
	} break;
	case MOMENT_OF_INERTIA_CAPSULE:
	{
		dMassSetCapsuleTotal(
			&mass,
			body->mass,
			roundf(body->moiParams[0]),
			body->moiParams[1],
			body->moiParams[2]);
	} break;
	case MOMENT_OF_INERTIA_DEFAULT:
	default:
	{
		real32 aabb[6] = {};
		dGeomGetAABB((dGeomID)body->spaceID, aabb);

		real32 radius = fmaxf(
			aabb[1] - aabb[0], fmaxf(aabb[3] - aabb[2], aabb[5] - aabb[4]))
			/ 2.0f;

		dMassSetSphereTotal(&mass, body->mass, radius);
	} break;
	}

	if (body->inertiaType != MOMENT_OF_INERTIA_USER)
	{
		dMassTranslate(
			&mass,
			body->centerOfMass.x,
			body->centerOfMass.y,
			body->centerOfMass.z);
	}

	if (body->gravity)
	{
		dBodySetGravityMode(body->bodyID, 1);
	}
	else
	{
		dBodySetGravityMode(body->bodyID, 0);
	}

	if (!body->defaultDamping)
	{
		dBodySetLinearDamping(body->bodyID, body->linearDamping);
		dBodySetAngularDamping(body->bodyID, body->angularDamping);

		dBodySetLinearDampingThreshold(
			body->bodyID,
			body->linearDampingThreshold);
		dBodySetAngularDampingThreshold(
			body->bodyID,
			body->angularDampingThreshold);
	}

	dBodySetMaxAngularSpeed(body->bodyID, body->maxAngularSpeed);

	if (body->dynamic)
	{
		dBodySetDynamic(body->bodyID);
		dBodySetMass(body->bodyID, &mass);
	}
	else
	{
		dBodySetKinematic(body->bodyID);
	}

	updateRigidBodyPosition(scene, coll, body, trans);

	if (body->enabled)
	{
		dBodyEnable(body->bodyID);
	}
	else
	{
		dBodyDisable(body->bodyID);
	}
}

void updateRigidBodyPosition(
	Scene *scene,
	CollisionComponent *coll,
	RigidBodyComponent *body,
	TransformComponent *trans)
{
	dBodySetAngularVel(
		body->bodyID,
		body->angularVel.x,
		body->angularVel.y,
		body->angularVel.z);

	dBodySetLinearVel(
		body->bodyID,
		body->velocity.x,
		body->velocity.y,
		body->velocity.z);

	dBodySetPosition(
		body->bodyID,
		trans->globalPosition.x,
		trans->globalPosition.y,
		trans->globalPosition.z);

	dQuaternion rot = {};
	rot[0] = trans->globalRotation.w;
	rot[1] = trans->globalRotation.x;
	rot[2] = trans->globalRotation.y;
	rot[3] = trans->globalRotation.z;

	dBodySetQuaternion(body->bodyID, rot);

	if (coll)
	{
		updateCollisionGeoms(scene, trans, coll);
	}
}

void destroyRigidBody(RigidBodyComponent *body)
{
	dSpaceDestroy(body->spaceID);
	dBodyDestroy(body->bodyID);
}

File: resources/scripts/components/rigid_body.lua
ffi.cdef[[
typedef enum moment_of_inertia_e
{
  MOMENT_OF_INERTIA_USER = -1,
  MOMENT_OF_INERTIA_DEFAULT = 0,
  MOMENT_OF_INERTIA_SPHERE = 1,
  MOMENT_OF_INERTIA_CUBE,
  MOMENT_OF_INERTIA_CAPSULE
} MomentOfInertia;

typedef struct rigid_body_component_t
{
  int64 bodyID;
  int64 spaceID;
  bool enabled;
  bool dynamic;
  bool gravity;
  real32 mass;
  kmVec3 centerOfMass;
  kmVec3 velocity;
  kmVec3 angularVel;
  bool defaultDamping;
  real32 linearDamping;
  real32 angularDamping;
  real32 linearDampingThreshold;
  real32 angularDampingThreshold;
  real32 maxAngularSpeed;
  MomentOfInertia inertiaType;
  real32 moiParams[6];
} RigidBodyComponent;
]]

local component = engine.components:register("rigid_body", "RigidBodyComponent")

local C = engine.C
local kazmath = engine.kazmath

function component:addVelocity(velocity)
  kazmath.kmVec3Add(self.velocity, self.velocity, velocity)
end

function component:addVelocityDirection(magnitude, direction)
  local velocity = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(velocity, direction, magnitude)

  self:addVelocity(velocity[0])
end

function component:addVelocityQuaternion(magnitude, quaternion)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:addVelocityDirection(magnitude, direction[0])
end

function component:addAngularVelocity(angularVelocity)
  kazmath.kmVec3Add(self.angularVel, self.angularVel, angularVelocity)
end

function component:addAngularVelocityAxis(magnitude, axis)
  local angularVelocity = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(angularVelocity, axis, magnitude)

  self:addAngularVelocity(angularVelocity[0])
end

function component:addAngularVelocityQuaternion(magnitude, quaternion)
  local axis = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(axis, quaternion)

  self:addAngularVelocityAxis(magnitude, axis[0])
end

function component:addAcceleration(acceleration, position)
  local force = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(force, acceleration, self.mass)

  local centerOfMass = ffi.new("kmVec3")
  position = position or centerOfMass

  self:addForce(force[0], position)
end

function component:addAccelerationDirection(magnitude, direction, position)
  local acceleration = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(acceleration, direction, magnitude)

  self:addAcceleration(acceleration[0], position)
end

function component:addAccelerationQuaternion(magnitude, quaternion, position)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:addAccelerationDirection(magnitude, direction[0], position)
end

function component:addForce(force, position)
  local centerOfMass = ffi.new("kmVec3")
  position = position or centerOfMass

  C.addForce(self, force, position)
end

function component:addForceDirection(magnitude, direction, position)
  local force = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(force, direction, magnitude)

  self:addForce(force[0], position)
end

function component:addForceQuaternion(magnitude, quaternion, position)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:addForceDirection(magnitude, direction[0], position)
end

function component:addTorque(torque)
  C.addTorque(self, torque)
end

function component:addTorqueAxis(magnitude, axis)
  local torque = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(torque, axis, magnitude)

  self:addTorque(torque[0])
end

function component:addTorqueQuaternion(magnitude, quaternion)
  local axis = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(axis, quaternion)

  self:addTorqueAxis(magnitude, axis[0])
end

function component:setVelocity(velocity)
  kazmath.kmVec3Assign(self.velocity, velocity)
end

function component:setVelocityDirection(magnitude, direction)
  local velocity = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(velocity, direction, magnitude)

  self:setVelocity(velocity[0])
end

function component:setVelocityQuaternion(magnitude, quaternion)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:setVelocityDirection(magnitude, direction[0])
end

function component:setAngularVelocity(angularVelocity)
  kazmath.kmVec3Assign(self.angularVel, angularVelocity)
end

function component:setAngularVelocityAxis(magnitude, axis)
  local angularVelocity = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(angularVelocity, axis, magnitude)

  self:setAngularVelocity(angularVelocity[0])
end

function component:setAngularVelocityQuaternion(magnitude, quaternion)
  local axis = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(axis, quaternion)

  self:setAngularVelocityAxis(magnitude, axis[0])
end

function component:setAcceleration(acceleration)
  local force = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(force, acceleration, self.mass)

  self:setForce(force[0])
end

function component:setAccelerationDirection(magnitude, direction)
  local acceleration = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(acceleration, direction, magnitude)

  self:setAcceleration(acceleration[0])
end

function component:setAccelerationQuaternion(magnitude, quaternion)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:setAccelerationDirection(magnitude, direction[0])
end

function component:setForce(force)
  C.setForce(self, force)
end

function component:setForceDirection(magnitude, direction, position)
  local force = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(force, direction, magnitude)

  self:setForce(force[0])
end

function component:setForceQuaternion(magnitude, quaternion, position)
  local direction = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(direction, quaternion)

  self:setForceDirection(magnitude, direction[0])
end

function component:setTorque(torque)
  C.setTorque(self, torque)
end

function component:setTorqueAxis(magnitude, axis)
  local torque = ffi.new("kmVec3[1]")
  kazmath.kmVec3Scale(torque, axis, magnitude)

  self:setTorque(torque[0])
end

function component:setTorqueQuaternion(magnitude, quaternion)
  local axis = ffi.new("kmVec3[1]")
  kazmath.kmQuaternionGetForwardVec3RH(axis, quaternion)

  self:setTorqueAxis(magnitude, axis[0])
end

function component:zeroVelocity()
  kazmath.kmVec3Zero(self.velocity)
end

function component:zeroAngularVelocity()
  kazmath.kmVec3Zero(self.angularVel)
end

function component:zeroForce()
  local force = ffi.new("kmVec3")
  self:setForce(force)
end

function component:zeroTorque()
  local torque = ffi.new("kmVec3")
  self:setTorque(torque)
end

function component:getVelocityMagnitude()
  return kazmath.kmVec3Length(self.velocity)
end

function component:getAngularVelocityMagnitude()
  return kazmath.kmVec3Length(self.angularVel)
end

function component:getAcceleration()
  local acceleration = ffi.new("kmVec3[1]")
  acceleration[0] = self:getForce()
  kazmath.kmVec3Scale(acceleration, acceleration[0], 1.0 / self.mass)
  return acceleration[0]
end

function component:getAccelerationMagnitude()
  local acceleration = self:getAcceleration()
  return kazmath.kmVec3Length(acceleration)
end

function component:getForce()
  return C.getForce(self)
end

function component:getForceMagnitude()
  local force = self:getForce()
  return kazmath.kmVec3Length(force)
end

function component:getTorque()
  return C.getTorque(self)
end

function component:getTorqueMagnitude()
  local torque = self:getTorque()
  return kazmath.kmVec3Length(torque)
end